libs/@guardian/libs/src/storage/storage.ts

```diff
@@ -98,6 +98,38 @@ class StorageFactory {
 	setRaw(key: string, value: string): void {
 		return this.#storage?.setItem(key, value);
 	}
+
+	/**
+	 * Get all of the keys in storage.
+	 *
+	 * @returns an array of all keys in the storage object,
+	 * or undefined if storage is not available.
+	 *
+	 * @note This is not a method in the native API, and gets
+	 * all keys by iterating over the entire storage object.
+	 * It is therefore potentially subject to race conditions
+	 * and if other code is modifying storage at the same time
+	 * then the results may be inaccurate.
+	 *
+	 * @note This method depends on the `key` method on the native API,
+	 * which gets keys by index. But the _order_ of keys in storage is
+	 * dependent on user-agent implementation, and so you should not rely
+	 * on the order of keys returned by this method. (See
+	 * https://developer.mozilla.org/en-US/docs/Web/API/Storage/key for more.)
+	 */
+	keys(): string[] | undefined {
+		if (!this.#storage) {
+			return undefined;
+		}
+		const allKeys: string[] = [];
+		for (let i = 0; i < this.#storage.length; i++) {
+			const maybeKey = this.#storage.key(i);
+			if (maybeKey) {
+				allKeys.push(maybeKey);
+			}
+		}
+		return allKeys;
```

Comment:
Unfortunately storage isn't iterable. As far as I'm away the `key` method is the only supported method for getting keys out of it?

Reply:
Just for your information, if you want to avoid mutation you can create an array of a specific length like so:

```ts
return Array.from({ length: this.#storage.length }, (_, i) =>
    this.#storage?.key(i),
    ).filter(isNonNullable);
```